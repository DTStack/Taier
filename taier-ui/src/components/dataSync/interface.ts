import type { DATA_SOURCE_ENUM } from '@/constant';

/**
 * 前端表单保存的值
 */
export interface ISourceFormField {
	sourceId?: number;
	table?: string | string[];
	/**
	 * Only used in Oracle and PostgreSQL
	 */
	schema?: string;
	where?: string;
	splitPK?: string;
	extralConfig?: string;
	increColumn?: string | number;
	/**
	 * Only used in HDFS
	 */
	path?: string;
	/**
	 * Only used in HDFS
	 */
	fileType?: 'orc' | 'text' | 'parquet';
	/**
	 * Only used in HDFS
	 */
	fieldDelimiter?: string;
	/**
	 * Only used in HDFS
	 */
	encoding?: 'utf-8' | 'gbk';
	/**
	 * Only used in Hive and SparkShrift
	 */
	partition?: string;
}

export interface ITargetFormField {
	sourceId?: number;
	table?: string;
	preSql?: string;
	postSql?: string;
	schema?: string;
	extralConfig?: string;
	partition?: string;
	path?: string;
	fileName?: string;
	fileType?: 'orc' | 'text' | 'parquet';
	fieldDelimiter?: string;
	encoding?: 'utf-8' | 'gbk';
	writeMode?: 'NONCONFLICT' | 'APPEND' | 'insert' | 'replace' | 'update';
}

/**
 * 传给后端的值
 */
export interface ISourceMapProps {
	name?: string;
	sourceId?: number;
	column?: IDataColumnsProps[];
	increColumn?: string | number;
	sourceList?: {
		key: string;
		tables?: string[] | string;
		type: DATA_SOURCE_ENUM;
		name: string;
		sourceId?: number;
	}[];
	extralConfig?: string;
	schema?: string;
	type?: Omit<ISourceFormField, 'sourceId'> & { type?: DATA_SOURCE_ENUM };

	[key: string]: any;
}

export interface ITargetMapProps {
	sourceId?: number;
	name?: string;
	extralConfig?: string;
	column?: IDataColumnsProps[];
	type?: Omit<ITargetFormField, 'sourceId'> & { type?: DATA_SOURCE_ENUM; rowkey?: string };

	[key: string]: any;
}

export interface IKeyMapProps {
	source: IDataColumnsProps[];
	target: IDataColumnsProps[];
}

export interface IDataColumnsProps {
	comment?: string;
	isPart?: boolean;
	key: string | number;
	part?: boolean;
	type: string;
	value?: string;
	index?: string;
	cf?: string;
	format?: string;
}

export interface IChannelFormProps {
	speed: string;
	channel: string;
	record?: number;
	percentage?: number;
	isRestore?: boolean;
	isSaveDirty?: boolean;
	tableName?: string;
	lifeDay?: string | number;
	restoreColumnName?: string | number;
}
