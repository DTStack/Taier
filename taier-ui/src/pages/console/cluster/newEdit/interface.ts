import type {
	COMPONENT_TYPE_VALUE,
	ENGINE_SOURCE_TYPE_ENUM,
	FLINK_DEPLOY_NAME,
} from '@/constant';

export interface IComponentProps {
	componentConfig: string;
	componentName: string;
	componentTemplate: string;
	componentTypeCode: number;
	deployType: keyof typeof FLINK_DEPLOY_NAME;
	engineId: ENGINE_SOURCE_TYPE_ENUM;
	gmtCreate: number;
	gmtModified: number;
	versionName: string;
	id: number;
	isDefault: boolean;
	isMetadata?: number;
	storeType: number;
	uploadFileName: string;
	kerberosFileName: string;
	mergeKrb5Content: string;
	paramsFile?: string;
	principal?: string;
	principals?: string;
}

export interface IScheduleComponentComp {
	componentTypeCode: COMPONENT_TYPE_VALUE;
	multiVersion: (IComponentProps | undefined)[];
}

export interface IScheduleComponent {
	components: IScheduleComponentComp[];
	schedulingCode: number;
	schedulingName: string;
}

export interface ISaveCompsData {
	key: number;
	value: string;
}

export interface IModifyComp {
	typeCode: COMPONENT_TYPE_VALUE;
	versionName?: string;
}

export interface IEditClusterRefProps {
	testConnects: (params?: ITestConnectsParams, callBack?: (bool: boolean) => void) => void;
	handleComplete: () => void;
}

export type IConfirmComps = IScheduleComponentComp | IComponentProps | COMPONENT_TYPE_VALUE[];

export interface IGetLoadTemplateParams {
	compVersion?: string;
	storeType?: number;
	deployType?: number;
}

export interface ITestConnectsParams {
	versionName?: string;
	typeCode?: COMPONENT_TYPE_VALUE;
	deployType?: number;
}

export interface IClusterInfo {
	clusterId: string;
	clusterName: string;
}

export interface ITestErrorMsg {
	componentVersion?: string | null;
	errorMsg: string | null;
}

export interface ITestStatusProps {
	clusterResourceDescription?: null | string;
	componentTypeCode: COMPONENT_TYPE_VALUE;
	componentVersion?: string | null;
	multiVersion?: ITestStatusProps[];
	errorMsg: string | null | ITestErrorMsg[];
	result: null | boolean;
}

export interface IVersionProps {
	dependencyKey: string | null;
	dependencyValue: string | null;
	deployTypes: number[] | null;
	id: number;
	key: string;
	required: boolean;
	type: string | null;
	value: string;
	values: IVersionProps[] | null;
}

export type IVersionData = Record<string, IVersionProps[]>;

export interface ICompTemplate {
	dependencyKey?: string;
	dependencyValue?: string;
	key: string;
	required: boolean;
	type: string;
	value: string | string[];
	values?: ICompTemplate[];
	isSameKey?: boolean;
}

export type ISaveComp = (params: Partial<IComponentProps>, type?: string) => void;

export type IHandleConfirm = (action: string, comps: IConfirmComps, mulitple?: boolean) => void;

export type ITestConnects = (params?: ITestConnectsParams, callBack?: (bool: boolean) => void) => void;

export type IHandleCompVersion = (typeCode: number, version: string) => void;
