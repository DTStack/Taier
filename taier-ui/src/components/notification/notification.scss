@import "@/styles/variables.scss";

.dt-notification.ant-notification-notice {
    padding: 5px 13px;
    cursor: pointer;

    .ant-notification-notice-message {
        margin-bottom: 0;
        overflow: hidden;
        font-size: $font-size;
        white-space: nowrap;
        text-overflow: ellipsis;
    }

    .ant-notification-notice-close {
        top: 6px;
        right: 14px;
    }

    &:hover {
        background-color: var(--button-hoverBackground);
    }
}
