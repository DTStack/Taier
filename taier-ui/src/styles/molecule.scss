@import "./variables.scss";

.mo-menuBar--horizontal {
    height: $menuBar-size;
    background: var(--activityBar-background);

    > .mo-menu {
        height: $menuBar-size;
        color: var(--activityBar-inactiveForeground);
        background: inherit;
    }
}

// only the tabs in editor should have hover styles
.mo-editor__group {
    .mo-tab__item:hover {
        background-color: var(--tab-hoverBackground);
    }
}

.mo-tab__item__status,
.mo-tab__item__op {
    color: inherit;

    &:hover {
        color: inherit;
    }
}

.mo-editor__group-breadcrumb {
    color: var(--breadcrumb-foreground);

    .mo-breadcrumb__item:hover {
        color: var(--breadcrumb-focusForeground);
        opacity: 1;
    }
}

.mo-notification__item:focus {
    background: var(--list-hoverBackground);
    outline: 1px solid var(--list-focusOutline);
}

.mo-statusBar__item > a:empty {
    padding: 0;
}

// ===================== Auxiliary Bar ===========================
.mo-auxiliaryBar__tabs {
    .mo-auxiliaryBar__tabs__tab {
        font-size: 12px;
        padding-top: 5px;
        padding-bottom: 5px;

        &:hover {
            color: var(--taier-rightBar-foreground);
            background-color: var(--taier-rightBar-background);
        }

        &--active {
            color: var(--taier-rightBar-foreground);
            background-color: var(--taier-rightBar-background);
            position: relative;

            &::after {
                position: absolute;
                top: 0;
                right: 0;
                bottom: 0;
                width: 2px;
                height: 100%;
                background-color: #3f87ff;
                content: "";
            }
        }
    }
}

.mo-auxiliaryBar__container {
    background: var(--sideBar-background);
}
