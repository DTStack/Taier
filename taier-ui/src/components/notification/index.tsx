import { CheckCircleOutlined, CloseCircleOutlined } from '@ant-design/icons';
import molecule from '@dtinsight/molecule';
import { notification as antNotification } from 'antd';
import type { ArgsProps } from 'antd/lib/notification';
import './notification.scss';

type INotificationConfigs = Pick<ArgsProps, 'message'> & { key: string };

class Notification {
    private assertNotExistNotification(key: string) {
        const { data } = molecule.notification.getState();
        return !data?.find((n) => n.id === key);
    }
    private showNotification() {
        const { showNotifications } = molecule.notification.getState();
        if (!showNotifications) {
            molecule.notification.toggleNotification();
        }
    }
    private highlightNotification(key: string) {
        const notificationItem = document.getElementById(key)?.parentElement;
        if (notificationItem) {
            notificationItem.tabIndex = 0;
            notificationItem.focus();
        }
    }
    /**
     * Add a notification both in antd and molecule
     */
    openWithMolecule(config: INotificationConfigs) {
        const { showNotifications } = molecule.notification.getState();
        // the antd's notification pops up only when the molecule's notification invisible
        if (!showNotifications) {
            antNotification.open({
                key: config.key,
                message: config.message,
                placement: 'bottomRight',
                className: 'dt-notification',
                duration: 8,
                closeIcon: <></>,
                onClose() {},
                onClick: () => {
                    antNotification.close(config.key);
                    this.showNotification();
                    setTimeout(() => {
                        this.highlightNotification(config.key);
                    }, 0);
                },
            });
        }

        if (this.assertNotExistNotification(config.key)) {
            molecule.notification.add([
                {
                    id: config.key,
                    value: '',
                    render() {
                        return config.message;
                    },
                },
            ]);
        }
    }

    /**
     * Open a notification both in antd and molecule in bottomRight with danger icon
     */
    error({ key, message }: { key: string; message: string }) {
        this.openWithMolecule({
            key,
            message: (
                <>
                    <CloseCircleOutlined
                        style={{
                            color: 'var(--editorError-foreground)',
                            marginRight: 5,
                        }}
                    />
                    <span id={key} title={message}>
                        {message}
                    </span>
                </>
            ),
        });
    }

    success({ key, message }: { key: string; message: string }) {
        this.openWithMolecule({
            key,
            message: (
                <>
                    <CheckCircleOutlined
                        style={{
                            color: 'var(--terminal-ansiBrightGreen)',
                            marginRight: 5,
                        }}
                    />
                    <span id={key} title={message}>
                        {message}
                    </span>
                </>
            ),
        });
    }
}
export default new Notification();
