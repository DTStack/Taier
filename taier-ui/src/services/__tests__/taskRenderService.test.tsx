import api from '@/api';
import notification from '@/components/notification';
import { TASK_TYPE_ENUM } from '@/constant';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { Form, Modal } from 'antd';
import React from 'react';
import TaskRenderService from '../taskRenderService';
import '@testing-library/jest-dom';
import molecule from '@dtinsight/molecule';
import type { IOfflineTaskProps } from '@/interface';
import { RightBarKind } from '@/interface';
import { editorActionBarService } from '..';

jest.mock('@/api');
jest.mock('@/api/http');
jest.mock('@/components/notification');

jest.mock('@/services', () => ({
    breadcrumbService: { getBreadcrumb: jest.fn() },
    editorActionBarService: { performSyncTaskActions: jest.fn() },
}));

jest.mock('@/components/scaffolds/create', () => ({
    datasource: ({ disabled, onChange }: any) => (
        <div data-testid="datasource" data-disabled={disabled} onClick={() => onChange()}>
            datasource
        </div>
    ),
    queue: () => <div>queue</div>,
}));

jest.mock('@/pages/editor/flink', () => () => {
    return <div>{`I'm flink`}</div>;
});

const mockSparkSQL = {
    key: 0,
    value: 'SparkSQL',
    computeType: 1,
    taskProperties: {
        renderCondition: null,
        renderKind: 'editor',
        formField: ['datasource', 'queue', 'test'],
        actionsCondition: null,
        actions: ['SAVE_TASK'],
        dataTypeCodes: [45],
        barItem: ['task', 'dependency', 'task_params', 'env_params'],
        barItemCondition: null,
    },
    jobType: 0,
};

const mockFlink = {
    key: 11,
    value: 'Flink',
    computeType: 0,
    taskProperties: {
        renderCondition: null,
        renderKind: 'flink',
        formField: ['resourceIdList', 'mainClass', 'exeArgs', 'componentVersion'],
        actionsCondition: null,
        actions: ['SAVE_TASK', 'SUBMIT_TASK', 'OPERATOR_TASK'],
        dataTypeCodes: null,
        barItem: ['env_params'],
        barItemCondition: null,
    },
    jobType: 1,
};

const mockSync = {
    key: 2,
    value: '数据同步',
    computeType: 1,
    taskProperties: {
        renderCondition: null,
        renderKind: 'dataSync1',
        formField: [],
        actionsCondition: {
            key: 'createModel',
            value: 0,
            actions: ['SAVE_TASK'],
        },
        actions: [],
        dataTypeCodes: null,
        barItem: [],
        barItemCondition: {
            barItem: ['task', 'env_params'],
            key: 'createModel',
            value: 0,
        },
    },
    jobType: 2,
};

beforeEach(() => {
    (api.getTaskTypes as jest.Mock).mockResolvedValue({
        code: 1,
        data: [mockSparkSQL, mockFlink, mockSync],
    });

    (api.getSupportSource as jest.Mock).mockResolvedValue({
        code: 1,
        data: {
            writers: [1],
            readers: [1],
            dataAcquisitionWriter: null,
            dataAcquisitionReader: null,
            flinkSqlSources: [26, 17, 37, 14],
            flinkSqlSinks: [1, 39, 33, 46],
            flinkSqlSides: [1],
        },
    });
});

describe('Test TaskRenderService', () => {
    it('Should filled with default value', async () => {
        const taskRenderService = new TaskRenderService();

        await waitFor(() => {
            expect(taskRenderService.getState().supportTaskList).toHaveLength(3);
            expect(taskRenderService.getState().supportSourceList.writers).toHaveLength(1);

            expect(taskRenderService.getState()).toMatchSnapshot();
        });

        expect(taskRenderService.getField(TASK_TYPE_ENUM.SPARK_SQL)).toEqual(mockSparkSQL);
        expect(taskRenderService.getRenderKind(TASK_TYPE_ENUM.SPARK_SQL)).toBe('editor');
    });

    it('Should get form field components', async () => {
        const taskRenderService = new TaskRenderService();
        const original = Modal.confirm;
        Modal.confirm = jest.fn();

        await waitFor(() => {
            expect(taskRenderService.getState().supportTaskList).toHaveLength(3);
            expect(taskRenderService.getState().supportSourceList.writers).toHaveLength(1);
        });

        // TASK_TYPE_ENUM.PY_SPARK does NOT have fields
        expect(taskRenderService.renderCreateForm(TASK_TYPE_ENUM.PY_SPARK)).toBe(null);

        // TASK_TYPE_ENUM.SPARK_SQL have ['datasource', 'queue', 'test'] field
        const { asFragment, getByTestId } = render(
            <Form>{taskRenderService.renderCreateForm(TASK_TYPE_ENUM.SPARK_SQL)!}</Form>
        );

        // 「test」 field is undefined so it should log error
        expect(notification.error).toBeCalledWith({
            key: 'UNDEFINED_DEFINED_COMPONENT',
            message: `未定义的表单组件-「test」`,
        });
        expect(asFragment()).toMatchSnapshot();

        const dataSourceEle = getByTestId('datasource');

        expect(dataSourceEle.dataset.disabled).toBe('false');

        fireEvent.click(dataSourceEle);
        expect(Modal.confirm).toBeCalledTimes(1);

        Modal.confirm = original;
    });

    it('Should all icons', () => {
        const taskRenderService = new TaskRenderService();

        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.SPARK_SQL)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.SYNC)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.HIVE_SQL)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.SQL)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.DATA_ACQUISITION)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.FLINK)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.OCEANBASE)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.VIRTUAL)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.WORK_FLOW)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.PYTHON)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.SHELL)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.CLICKHOUSE)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.SPARK)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.DORIS)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.MYSQL)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.GREENPLUM)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.POSTGRE_SQL)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.SQL_SERVER)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.TiDB)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(TASK_TYPE_ENUM.VERTICA)).toMatchSnapshot();
        expect(taskRenderService.renderTaskIcon(100 as TASK_TYPE_ENUM)).toMatchSnapshot();
    });

    it('Should render specific tab on Editor', async () => {
        const taskRenderService = new TaskRenderService();

        await waitFor(() => {
            expect(taskRenderService.getState().supportTaskList).toHaveLength(3);
            expect(taskRenderService.getState().supportSourceList.writers).toHaveLength(1);
        });

        // Common task like sparkSQL
        expect(
            taskRenderService.renderTabOnEditor(TASK_TYPE_ENUM.SPARK_SQL, {
                id: 'sparkSQL',
                name: 'sparkSQL',
                taskType: TASK_TYPE_ENUM.SPARK_SQL,
                sqlText: '-- test',
            })
        ).toMatchSnapshot();

        // FlinkSQL doesn't have value and language in data
        const flinkTab = taskRenderService.renderTabOnEditor(TASK_TYPE_ENUM.FLINK, {
            id: 'flink',
            name: 'flink',
            taskType: TASK_TYPE_ENUM.FLINK,
        });
        expect(flinkTab).toMatchSnapshot();
        const { findByText } = render(<>{(flinkTab.renderPane as (data: any) => React.ReactNode)({ id: 'test' })}</>);
        expect(await findByText("I'm flink")).toBeInTheDocument();

        // Sync have value in data
        expect(
            taskRenderService.renderTabOnEditor(TASK_TYPE_ENUM.SYNC, {
                id: 'sync',
                name: 'sync',
                taskType: TASK_TYPE_ENUM.SYNC,
                value: '{"a": 1}',
            })
        ).toMatchSnapshot();
    });

    it('Should render specific right bar', async () => {
        const taskRenderService = new TaskRenderService();

        await waitFor(() => {
            expect(taskRenderService.getState().supportTaskList).toHaveLength(3);
            expect(taskRenderService.getState().supportSourceList.writers).toHaveLength(1);
        });

        (molecule.editor.getState as jest.Mock)
            .mockReset()
            .mockImplementationOnce(() => ({
                current: {
                    tab: {
                        id: 'edit-datasource',
                    },
                },
            }))
            .mockImplementationOnce(() => ({
                current: {
                    tab: {
                        id: 1,
                        data: {
                            taskType: 0,
                        },
                    },
                },
            }))
            .mockImplementationOnce(() => ({
                current: {
                    tab: {
                        id: 1,
                        data: {
                            taskType: 2,
                            createModel: 0,
                        },
                    },
                },
            }))
            .mockImplementationOnce(() => ({
                current: {
                    tab: {
                        id: 1,
                        data: {
                            taskType: 2,
                            createModel: 1,
                        },
                    },
                },
            }));

        // invalid tab
        expect(taskRenderService.renderRightBar()).toEqual([RightBarKind.TASK]);

        // sparkSQL tab
        expect(taskRenderService.renderRightBar()).toEqual([
            RightBarKind.TASK,
            'dependency',
            'task_params',
            'env_params',
        ]);

        // sync tab with createModel(0)
        expect(taskRenderService.renderRightBar()).toEqual([RightBarKind.TASK, 'env_params']);
        // sync tab with createModel(1)
        expect(taskRenderService.renderRightBar()).toEqual([RightBarKind.TASK]);
    });

    it('Should renderEditorActions', async () => {
        const taskRenderService = new TaskRenderService();

        await waitFor(() => {
            expect(taskRenderService.getState().supportTaskList).toHaveLength(3);
            expect(taskRenderService.getState().supportSourceList.writers).toHaveLength(1);
        });

        expect(taskRenderService.renderEditorActions(TASK_TYPE_ENUM.SPARK_SQL, {} as IOfflineTaskProps)).toEqual([
            {
                icon: 'save',
                id: 'task.save',
                name: 'Save Task',
                place: 'outer',
                title: '保存',
            },
        ]);

        expect(
            taskRenderService.renderEditorActions(TASK_TYPE_ENUM.SYNC, {
                createModel: 0,
            } as IOfflineTaskProps)
        ).toEqual([
            {
                icon: 'save',
                id: 'task.save',
                name: 'Save Task',
                place: 'outer',
                title: '保存',
            },
        ]);

        expect(
            taskRenderService.renderEditorActions(TASK_TYPE_ENUM.SYNC, {
                createModel: 1,
            } as IOfflineTaskProps)
        ).toEqual([]);
    });

    it('Should focus tab already opened', async () => {
        (molecule.editor.isOpened as jest.Mock).mockReset().mockImplementation(() => true);
        (molecule.editor.getGroupIdByTab as jest.Mock).mockReset().mockImplementation(() => 1);
        (molecule.editor.setActive as jest.Mock).mockReset();
        (editorActionBarService.performSyncTaskActions as jest.Mock).mockReset();

        const taskRenderService = new TaskRenderService();
        taskRenderService.openTask({ id: 1, taskType: 0 });
        expect(molecule.editor.setActive).toBeCalledWith(1, '1');
        await waitFor(() => {
            // Don't forget perform actions when tab changed
            expect(editorActionBarService.performSyncTaskActions).toBeCalled();
        });
    });

    it('Should open tab after request', async () => {
        (molecule.editor.isOpened as jest.Mock).mockReset().mockImplementation(() => false);
        (api.getOfflineTaskByID as jest.Mock).mockReset().mockResolvedValue({
            code: 1,
            data: {
                id: 'sparkSQL',
                name: 'sparkSQL',
                taskType: TASK_TYPE_ENUM.SPARK_SQL,
                sqlText: '-- test',
            },
        });
        (molecule.folderTree.setActive as jest.Mock).mockReset();
        (molecule.editor.open as jest.Mock).mockReset();
        (molecule.explorer.forceUpdate as jest.Mock).mockReset();

        const taskRenderService = new TaskRenderService();
        taskRenderService.openTask({ id: 1, taskType: 0 });

        await waitFor(() => {
            expect(molecule.folderTree.setActive).toBeCalledWith('sparkSQL');
            expect(molecule.editor.open).toBeCalled();
            expect(molecule.explorer.forceUpdate).toBeCalled();
        });
    });

    it('Should open tab without request', () => {
        (molecule.editor.isOpened as jest.Mock).mockReset().mockImplementation(() => false);

        (notification.error as jest.Mock).mockReset();

        (molecule.folderTree.setActive as jest.Mock).mockReset();
        (molecule.editor.open as jest.Mock).mockReset();
        (molecule.explorer.forceUpdate as jest.Mock).mockReset();

        const taskRenderService = new TaskRenderService();
        taskRenderService.openTask({ id: 1 }, { create: true });

        expect(notification.error).toBeCalledWith({
            key: 'OPEN_TASK_ERROR',
            message: `无法打开一个未知任务类型的任务，当前任务的任务类型为 undefined`,
        });

        taskRenderService.openTask({ id: 1, taskType: 0 }, { create: true });

        expect(notification.error).toBeCalledWith({
            key: 'OPEN_TASK_ERROR',
            message: `无法打开一个未知任务名称任务`,
        });

        taskRenderService.openTask({ id: 1, taskType: 0, name: 'test' }, { create: true });
        expect(molecule.folderTree.setActive).toBeCalledTimes(1);
        expect(molecule.editor.open).toBeCalled();
        expect(molecule.explorer.forceUpdate).toBeCalled();
    });

    it("Should log error when getting task's types failed", async () => {
        (api.getTaskTypes as jest.Mock).mockReset().mockResolvedValue({ code: 0 });
        (notification.error as jest.Mock).mockReset();

        const taskRenderService = new TaskRenderService();

        await waitFor(() => {
            expect(notification.error).toBeCalledWith({
                key: 'FailedJob',
                message: `获取支持的类型失败，将无法创建新的任务！`,
            });
        });

        taskRenderService.getTaskTypes(true);

        await waitFor(() => {
            // ONLY called once since doesn't log error at second scene
            expect(notification.error).toBeCalledTimes(1);
        });
    });
});
