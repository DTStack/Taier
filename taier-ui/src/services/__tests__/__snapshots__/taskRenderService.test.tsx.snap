// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TaskRenderService Should all icons 1`] = `
<SparkSQLIcon
  style={
    Object {
      "color": "#519aba",
    }
  }
/>
`;

exports[`Test TaskRenderService Should all icons 2`] = `"sync"`;

exports[`Test TaskRenderService Should all icons 3`] = `
<HiveSQLIcon
  style={
    Object {
      "color": "#4291f0",
    }
  }
/>
`;

exports[`Test TaskRenderService Should all icons 4`] = `
<FlinkSQLIcon
  style={
    Object {
      "color": "#5655d8",
    }
  }
/>
`;

exports[`Test TaskRenderService Should all icons 5`] = `
<DataCollectionIcon
  style={
    Object {
      "color": "#3F87FF",
    }
  }
/>
`;

exports[`Test TaskRenderService Should all icons 6`] = `<FlinkIcon />`;

exports[`Test TaskRenderService Should all icons 7`] = `<OceanBaseIcon />`;

exports[`Test TaskRenderService Should all icons 8`] = `<VirtualIcon />`;

exports[`Test TaskRenderService Should all icons 9`] = `
<WorkflowIcon
  style={
    Object {
      "color": "#2491F7",
    }
  }
/>
`;

exports[`Test TaskRenderService Should all icons 10`] = `<PythonIcon />`;

exports[`Test TaskRenderService Should all icons 11`] = `<ShellIcon />`;

exports[`Test TaskRenderService Should all icons 12`] = `<ClickhouseIcon />`;

exports[`Test TaskRenderService Should all icons 13`] = `<SparkIcon />`;

exports[`Test TaskRenderService Should all icons 14`] = `<DorisIcon />`;

exports[`Test TaskRenderService Should all icons 15`] = `<MysqlIcon />`;

exports[`Test TaskRenderService Should all icons 16`] = `<GreenPlumIcon />`;

exports[`Test TaskRenderService Should all icons 17`] = `<PostgreSqlIcon />`;

exports[`Test TaskRenderService Should all icons 18`] = `
<SqlServerIcon
  style={
    Object {
      "color": "#bf4339",
    }
  }
/>
`;

exports[`Test TaskRenderService Should all icons 19`] = `
<TiDbIcon
  style={
    Object {
      "color": "#bf4339",
    }
  }
/>
`;

exports[`Test TaskRenderService Should all icons 20`] = `<VerticaIcon />`;

exports[`Test TaskRenderService Should all icons 21`] = `"file"`;

exports[`Test TaskRenderService Should filled with default value 1`] = `
Object {
  "supportSourceList": Object {
    "dataAcquisitionReader": null,
    "dataAcquisitionWriter": null,
    "flinkSqlSides": Array [
      1,
    ],
    "flinkSqlSinks": Array [
      1,
      39,
      33,
      46,
    ],
    "flinkSqlSources": Array [
      26,
      17,
      37,
      14,
    ],
    "readers": Array [
      1,
    ],
    "writers": Array [
      1,
    ],
  },
  "supportTaskList": Array [
    Object {
      "computeType": 1,
      "jobType": 0,
      "key": 0,
      "taskProperties": Object {
        "actions": Array [
          "SAVE_TASK",
        ],
        "actionsCondition": null,
        "barItem": Array [
          "task",
          "dependency",
          "task_params",
          "env_params",
        ],
        "barItemCondition": null,
        "dataTypeCodes": Array [
          45,
        ],
        "formField": Array [
          "datasource",
          "queue",
          "test",
        ],
        "renderCondition": null,
        "renderKind": "editor",
      },
      "value": "SparkSQL",
    },
    Object {
      "computeType": 0,
      "jobType": 1,
      "key": 11,
      "taskProperties": Object {
        "actions": Array [
          "SAVE_TASK",
          "SUBMIT_TASK",
          "OPERATOR_TASK",
        ],
        "actionsCondition": null,
        "barItem": Array [
          "env_params",
        ],
        "barItemCondition": null,
        "dataTypeCodes": null,
        "formField": Array [
          "resourceIdList",
          "mainClass",
          "exeArgs",
          "componentVersion",
        ],
        "renderCondition": null,
        "renderKind": "flink",
      },
      "value": "Flink",
    },
    Object {
      "computeType": 1,
      "jobType": 2,
      "key": 2,
      "taskProperties": Object {
        "actions": Array [],
        "actionsCondition": Object {
          "actions": Array [
            "SAVE_TASK",
          ],
          "key": "createModel",
          "value": 0,
        },
        "barItem": Array [],
        "barItemCondition": Object {
          "barItem": Array [
            "task",
            "env_params",
          ],
          "key": "createModel",
          "value": 0,
        },
        "dataTypeCodes": null,
        "formField": Array [],
        "renderCondition": null,
        "renderKind": "dataSync1",
      },
      "value": "数据同步",
    },
  ],
}
`;

exports[`Test TaskRenderService Should get form field components 1`] = `
<DocumentFragment>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      data-disabled="false"
      data-testid="datasource"
    >
      datasource
    </div>
    <div>
      queue
    </div>
  </form>
</DocumentFragment>
`;

exports[`Test TaskRenderService Should render specific tab on Editor 1`] = `
Object {
  "breadcrumb": undefined,
  "data": Object {
    "id": "sparkSQL",
    "language": "sparksql",
    "name": "sparkSQL",
    "sqlText": "-- test",
    "taskType": 0,
    "value": "-- test",
  },
  "icon": <SparkSQLIcon
    style={
      Object {
        "color": "#519aba",
      }
    }
  />,
  "id": "sparkSQL",
  "name": "sparkSQL",
}
`;

exports[`Test TaskRenderService Should render specific tab on Editor 2`] = `
Object {
  "breadcrumb": undefined,
  "data": Object {
    "id": "flink",
    "name": "flink",
    "taskType": 11,
  },
  "icon": <FlinkIcon />,
  "id": "flink",
  "name": "flink",
  "renderPane": [Function],
}
`;

exports[`Test TaskRenderService Should render specific tab on Editor 3`] = `
Object {
  "breadcrumb": undefined,
  "data": Object {
    "id": "sync",
    "name": "sync",
    "taskType": 2,
    "value": undefined,
  },
  "icon": "sync",
  "id": "sync",
  "name": "sync",
  "renderPane": [Function],
}
`;
